doc/tutorials/sgx_virtualization.rst

```diff
@@ -0,0 +1,269 @@
+select code and data in order to prevent direct attacks on executing code or
+data stored in memory. SGX allows an application to instantiate a protected
+container, referred to as an enclave, which is protected against external
+software access, including privileged malware.
+
+
+High Level ACRN SGX Virtualization Design
+*****************************************
+
+ACRN SGX virtualization support can be divided into three parts:
+
+* SGX capability exposed to Guest EPC (Enclave Page Cache) management Enclave
+* System function handing
+* Enclave System function handling
+
+The image below shows the high-level design of SGX virtualization in ACRN.
+
+.. figure:: images/sgx-1.png
+   :width: 500px
+   :align: center
+
+   SGX Virtualization in ACRN
+
+
+Enable SGX support for Guest
+****************************
+
+Presumptions
+============
+
+No Enclave in a Hypervisor
+--------------------------
+
+ACRN does not support running an enclave in a hypervisor since the whole
+hypervisor is currently unning in VMX root mode, ring 0, and an enclave must
+run in ring 3. SGX virtualization in ACRN provides the SGX capability to (non-
+SOS) VMs.
+
+Enable SGX on Host
+------------------
+
+For SGX virtualization support in ACRN, you must manually enable the SGX
+feature and configure the Processor Reserved Memory (PRM) in the platform
+BIOS. (ACRN does not support the "Software Control" option to enable SGX at
+run time.) If SGX is not enabled or the hardware platform does not support
+SGX, ACRN SGX virtualization will not be enabled.
+
+EPC Page Swapping in Guest
+--------------------------
+
+ACRN only partitions the physical EPC resources for VMs. The Guest OS kernel
+handles EPC page swapping inside Guest.
+
+Instructions
+============
+
+SGX support for a Guest OS is not enabled by default. Follow these steps to
+enable SGX support in the BIOS and in ACRN:
+
+#. Check the system BIOS on your target platform to see if Intel SGX is
+   supported (CPUID.07H.EBX[2] should be 1).
+#. Enable the SGX feature in the BIOS setup screens. Follow these instructions:
+
+   a) Go to the Security page:
+
+      .. figure:: images/sgx-2.jpg
+         :width: 500px
+         :align: center
+
+   #) Enable SGX and configure the SGX Reserved Memory size as below:
+
+      * Intel Software Guard Extension (SGX) -> Enabled
+      * SGX Reserved Memory Size -> 128MB
+
+      .. figure:: images/sgx-3.jpg
+         :width: 500px
+         :align: center
+
+      .. note::
+
+         Not all SGX Reserved Memory can be used as EPC. On KBL-NUC-i7,
+         the SGX EPC size is 0x5d80000 (93.5MB) when the SGX Reserved Memory
+         Size is set to 128MB.
+
+#. Add the EPC config in the VM configuration:
+   Apply the patch to enable SGX support in UOS in the SDC scenario:
+
+   .. code-block:: bash
+
+      $ cd <projectacrn base folder>
+      $ curl https://github.com/binbinwu1/acrn-hypervisor/commit/
```

Comment:
This command has to be on one line

Reply:
Changed to:
      $ cd <projectacrn base folder>
      $ curl https://github.com/binbinwu1/acrn-hypervisor/commit \ 
          /0153b2b9b9920b61780163f19c6f5318562215ef.patch | git apply